### PR TITLE
Feat/api refactor

### PR DESCRIPTION
**Pull request by swaroop-osec**
- The current implementation always returns "not verified" when the signer is None because no data with the signer as None was found in the database. This happens because we enforce verification of the build only if a PDA for that program is found.
- Refactored the majority of the codebase: Improved code organization and optimized Functions and Logic.
- Updated inline documentation and README

### CODE REVIEW

**Comment by Woody4618** [comment]
This looks great! Just program authority will be null if a program get frozen. That may become a problem.

File: api/src/api/handlers/sync_verify.rs [hunk 1]
```diff
@@ -1,29 +1,83 @@
...
-        tracing::error!("Error inserting into database: {:?}", e);
-        return (
+}
+
+/// Checks if the program is already verified
+async fn check_for_duplicate(
+    payload: &SolanaProgramBuildParams,
+    db: &DbClient,
+) -> Option<(StatusCode, Json<ApiResponse>)> {
+    match check_and_handle_duplicates(payload, DEFAULT_SIGNER.to_string(), db).await {
```
[the remark this answers is not in the document]
Comment (swaroop-osec) [reply api/src/api/handlers/sync_verify.rs:61]: Ideally this should follow a similar structure to asynchronous verification. 
```
1 -> Get Otter PDA
2 -> Check for duplicates
3 -> Continue verification if no duplicates are found
```

I am pushing some changes for this.

File: api/src/api/handlers/sync_verify.rs [hunk 2]
```diff
@@ -1,29 +1,83 @@
...
+    );
+
+    // Create build record
     let verify_build_data = SolanaProgramBuild::from(&payload);
     let uuid = verify_build_data.id.clone();
 
-    // First check if the program is already verified
-    let is_dublicate = check_and_handle_duplicates(&payload, DEFAULT_SIGNER.to_string(), &db).await;
+    // Check for existing verification
+    if let Some(response) = check_for_duplicate(&payload, &db).await {
```
[the remark this answers is not in the document]
Comment (swaroop-osec) [reply api/src/api/handlers/sync_verify.rs:40]: I am updating it to send the signer dynamically from the on-chain data we received.

File: api/src/api/handlers/job_status.rs [hunk 3]
```diff
@@ -1,73 +1,103 @@
...
+    match db.get_job(&job_id).await {
+        Ok(job) => {
+            let status: JobStatus = job.status.into();
+            match status {
+                JobStatus::Completed => {
+                    info!("Job {} completed, fetching verification details", job_id);
+                    match db.get_verified_build(&job.program_id, None).await {
+                        Ok(verified_build) => {
+                            let repo_url = job.commit_hash.map_or(job.repository.clone(), |hash| {
+                                format!("{}/tree/{}", job.repository.trim_end_matches('/'), hash)
```
[the remark this answers is not in the document]
Comment (swaroop-osec) [reply api/src/api/handlers/job_status.rs:34]: This is the only place where we did this explicitly. We use `build_repository_url` in all other instances.

File: api/src/api/handlers/verification_status.rs [hunk 4]
```diff
@@ -4,31 +4,52 @@ use crate::db::models::{
...
-                on_chain_hash: result.on_chain_hash,
-                last_verified_at: result.last_verified_at,
-                executable_hash: result.executable_hash,
-                repo_url: result.repo_url,
-                commit: result.commit,
-            }
-            .into(),
-        ),
+        Ok(result) => {
+            let status_message = if result.is_verified {
```
[the remark this answers is not in the document]
Comment (swaroop-osec) [reply api/src/api/handlers/verification_status.rs:27]: That’s already implemented here [api/src/db/verification.rs#L49](https://github.com/otter-sec/solana-verified-programs-api/blob/d6bb51bc91d8f645cd6bdf3885a9666bae8730f0/api/src/db/verification.rs#L49) 
If the hash is not found in the Redis cache, we fetch the on-chain program hash.

File: api/src/api/handlers/verified_programs_list.rs [hunk 5]
```diff
@@ -1,21 +1,48 @@
...
+/// On failure, still returns an empty list but logs the error
 pub(crate) async fn get_verified_programs_list(
     State(db): State<DbClient>,
 ) -> (StatusCode, Json<VerifiedProgramListResponse>) {
-    let verified_programs = db.get_verified_programs().await.unwrap();
+    info!("Fetching list of verified programs");
 
-    // get all program ids from the verified_programs
+    let verified_programs = match db.get_verified_programs().await {
+        Ok(programs) => {
```
[the remark this answers is not in the document]
Comment (swaroop-osec) [reply api/src/api/handlers/verified_programs_list.rs:20]: That makes sense. Paginated results would be helpful. I’ll implement this as well.

File: api/src/api/init.rs [hunk 6]
```diff
@@ -69,56 +63,33 @@ pub fn initialize_router(db: DbClient) -> Router {
...
+        // Verification routes (stricter rate limits)
         .route("/verify", post(process_async_verification))
         .route(
             "/verify-with-signer",
             post(process_async_verification_with_signer),
         )
         .route("/verify_sync", post(process_sync_verification))
         .layer(
-            global_rate_limit(1)
+            global_rate_limit(5)
```
[the remark this answers is not in the document]
Comment (swaroop-osec) [reply api/src/api/init.rs:76]: Are you suggesting making this configurable via an environment variable or a constant? Could you provide more details on this?

File: api/src/api/init.rs [hunk 7]
```diff
@@ -69,56 +63,33 @@ pub fn initialize_router(db: DbClient) -> Router {
...
-        .layer(
-            global_rate_limit(10000)
-                .layer(rate_limit_per_ip(1, 100))
-                .layer(cors(Method::GET))
-                .layer(CompressionLayer::new().zstd(true)),
-        )
         .route("/unverify", post(handle_unverify))
         .layer(
-            global_rate_limit(10000)
+            global_rate_limit(100)
```
[the remark this answers is not in the document]
Comment (swaroop-osec) [reply api/src/api/init.rs:83]: No, only authorized users can unverify programs. You can see the authorization logic here: [unverify.rs#L70](https://github.com/otter-sec/solana-verified-programs-api/blob/fd6bb3a736d67dac40adeb453f8e9dbdd5088135/api/src/api/handlers/unverify.rs#L70)

File: api/src/db/programs.rs [hunk 8]
```diff
@@ -1,16 +1,82 @@
...
+
+    pub async fn get_verification_status_all(&self) -> Result<Vec<VerifiedProgramStatusResponse>> {
+
+        let all_verified_programs = self.get_verified_programs().await?;
+
+        let mut programs_status_all = Vec::new();
+
+        for program in all_verified_programs {
+        info!("Checking verification status for program: {}", program.program_id);
+        match self.clone().check_is_verified(program.program_id.to_owned(), None).await {
```
[the remark this answers is not in the document]
Comment (swaroop-osec) [reply api/src/db/programs.rs:35]: Currently, we search for any one of these signers (DEFAULT_SIGNER, SIGNER_KEYS, program_authority) and return the first one. However, I agree that we need a more robust selection process. You check the current implementation here: [verification.rs#L200](https://github.com/otter-sec/solana-verified-programs-api/blob/fd6bb3a736d67dac40adeb453f8e9dbdd5088135/api/src/db/verification.rs#L200).

File: api/src/services/verification.rs [hunk 9]
```diff
@@ -18,185 +28,197 @@ pub async fn check_and_process_verification(
...
-                {
+                match db.get_verified_build(&response.program_id, response.signer).await {
                     Ok(verified_build) => Some(VerifyResponse {
                         status: JobStatus::Completed,
                         request_id: verified_build.solana_build_id,
                         message: "Verification already completed.".to_string(),
                     }),
                     Err(err) => {
-                        tracing::error!("Error getting verified build: {:?}", err);
+                        error!("Failed to get verified build: {:?}", err);
```
[the remark this answers is not in the document]
Comment (swaroop-osec) [reply api/src/services/verification.rs:73]: The error is just being logged, we are returning None from this function.

File: api/src/api/init.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Woody4618) [reply api/src/api/init.rs:76]: Yes i though some env variable. But maybe doesnt matter.

**Pull request merged**